File: generated_code/GeminiPro/Java/017_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree that stores integers.
 * This class functions as an integer set, disallowing duplicate values.
 * The core feature is the splay operation, which moves an accessed element
 * to the root of the tree to optimize for future accesses.
 */
public class SplayTree {

    /**
     * Represents a node within the Splay Tree.
     * It is a private static inner class as it's tightly coupled with SplayTree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        /**
         * Constructs a new node with the given integer key.
         *
         * @param key The integer value for this node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Rotation Helper Methods ---

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * This is a helper method for the splay operation.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * This is a helper method for the splay operation.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    // --- Core Splay Operation ---

    /**
     * Performs the top-down splay operation on the given key.
     * This operation moves the node with the specified key to the root of the tree.
     * If the key is not found, the last accessed node on the search path is
     * moved to the root instead.
     *
     * @param key The key to splay towards the root.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Use a dummy node to simplify linking of left and right subtrees.
        Node header = new Node(0);
        Node leftTreeMax = header;
        Node rightTreeMin = header;

        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig case (Right-Right)
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link current to the right tree
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig case (Left-Left)
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link current to the left tree
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = header.right;
        current.right = header.left;

        root = current;
    }

    // --- Public API Methods ---

    /**
     * Inserts a key into the Splay Tree.
     * If the key already exists, the tree is splayed on that key, but no new
     * node is added, as this is an integer set.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        splay(key);

        // If the key is already at the root, it exists in the set.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     * If the key is found, it is removed, and the tree is restructured.
     * If the key is not found, the tree is splayed on the last accessed node,
     * but no deletion occurs.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        splay(key);

        // If the key is not at the root after splaying, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // The new root will be the largest element in the left subtree.
            // Temporarily make the left subtree the main tree.
            root = leftSubtree;
            // Splay for the (deleted) key. Since this key is larger than any
            // key in the new tree, this will bring the max element to the root.
            splay(key);
            // After splaying, the new root has no right child.
            // Attach the original right subtree here.
            root.right = rightSubtree;
        }
    }

    /**
     * Searches for a key in the tree and performs the splay operation.
     * This method moves the found node to the root. If the node is not found,
     * the last accessed node in the search path is moved to the root.
     *
     * @param key The key to search for.
     * @return The key if it is found; otherwise, returns null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        // After splaying, the key, if it exists, will be at the root.
        if (root.key == key) {
            return key;
        }
        return null;
    }
}

File: generated_code/GeminiPro/Python/017_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Splay Tree class that implements a
    dictionary-like set for integers.

    This implementation includes insert, delete, and search methods. The core
    feature of a splay tree is its self-optimizing nature: frequently accessed
    nodes are moved closer to the root, leading to amortized logarithmic time
    complexity for its operations.
    """

    class _Node:
        """A private helper class for the nodes of the splay tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on the subtree rooted at node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on the subtree rooted at node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on a node, moving it to the root.
        Splaying involves a sequence of rotations (Zig, Zig-Zig, Zig-Zag)
        until the node is the root of the tree.
        """
        if not node:
            return
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig step
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (node is left child, parent is right child)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        Performs the splaying operation on the accessed node if the key is found,
        or on the last accessed node (the parent of the potential location) if
        the key is not found.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise. After the operation,
            the splayed node will be the new root of the tree.
        """
        if not self.root:
            return False

        current = self.root
        last_node = self.root
        while current:
            last_node = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Key not found, splay the last accessed node to the root.
        self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then the new node is splayed
        to the root, becoming the new root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it to the root and we're done.
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Insert the new node as a child of the last-visited node.
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the newly inserted node to the root.
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The node containing the key is first splayed to the root and then removed.
        The tree is then restructured by joining the remaining left and right
        subtrees. If the key is not found, the last-accessed node in the search
        path is splayed to the root.

        Args:
            key: The integer key to delete.
        """
        # Splay the node with the given key (or its would-be parent) to the root.
        if not self.search(key):
            # Key was not found. search() has already splayed the last-accessed node.
            return

        # After search(), if the key was found, it is now at the root.
        # This check is slightly redundant due to the search above, but is safe.
        if not self.root or self.root.key != key:
            return

        # The node to delete is at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # If there is a left subtree, find its maximum element.
            # This maximum element will become the new root of the combined tree.
            left_subtree.parent = None  # Temporarily detach the left subtree.
            
            # Find the maximum node in the (now separate) left subtree.
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right
            
            # Splay this maximum node to the root of the left subtree.
            # We temporarily set self.root to perform the splay within this subtree.
            self.root = left_subtree
            self._splay(max_in_left)  # After this, self.root is max_in_left.
            
            # The new root's right child is the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root